File: typescript/ts_sdk/src/wns/resolver.ts
import {
  HandleGoneError,
  HandleMovedError,
  HandleNotFoundError,
  HandleResolutionError,
} from '../errors/index.js';
import type { DIDSubjectProfile, HandleResolutionDocument, ResolveHandleOptions } from './types.js';
import { HandleStatus, SubjectType } from './types.js';
import { buildResolutionUrl, parseWbaUri, validateHandle } from './validator.js';

export async function resolveHandle(
  handle: string,
  options: ResolveHandleOptions = {}
): Promise<HandleResolutionDocument> {
  const bareHandle = stripWbaScheme(handle);
  const [localPart, domain] = validateHandle(bareHandle);
  const normalized = `${localPart}.${domain}`;
  const baseUrl = options.baseUrlOverride?.replace(/\/$/, '');
  const url = baseUrl
    ? `${baseUrl}/.well-known/handle/${localPart}`
    : buildResolutionUrl(localPart, domain);

  void options.verifySsl;
  const timeoutMs = Math.round((options.timeoutSeconds ?? 10) * 1000);
  const response = await fetch(url, {
    headers: { Accept: 'application/json' },
    redirect: 'manual',
    signal: AbortSignal.timeout(timeoutMs),
  }).catch((error) => {
    throw new HandleResolutionError(
      `Network error resolving handle '${normalized}': ${(error as Error).message}`,
      502,
      error as Error
    );
  });

  if (response.status === 301) {
    throw new HandleMovedError(
      `Handle '${normalized}' has been migrated`,
      response.headers.get('Location') ?? ''
    );
  }
  if (response.status === 404) {
    throw new HandleNotFoundError(`Handle '${normalized}' does not exist`);
  }
  if (response.status === 410) {
    throw new HandleGoneError(`Handle '${normalized}' has been permanently revoked`);
  }
  if (!response.ok) {
    throw new HandleResolutionError(
      `Unexpected status ${response.status} resolving '${normalized}'`,
      502
    );
  }

  const payload = (await response.json()) as Record<string, unknown>;
  const document: HandleResolutionDocument = {
    handle: String(payload.handle ?? ''),
    did: String(payload.did ?? ''),
    status: normalizeStatus(String(payload.status ?? '')),
    updated: payload.updated ? String(payload.updated) : undefined,
    versionId: payload.versionId ? String(payload.versionId) : undefined,
    ttl: typeof payload.ttl === 'number' ? payload.ttl : undefined,
    profile: normalizeProfile(payload.profile),
  };
  if (document.handle.toLowerCase() !== normalized) {
    throw new HandleResolutionError(
      `Handle mismatch: requested '${normalized}', got '${document.handle}'`,
      502
    );
  }
  dropInvalidProfileProjection(document);
  return document;
}

export async function resolveHandleFromUri(
  wbaUri: string,
  options: ResolveHandleOptions = {}
): Promise<HandleResolutionDocument> {
  const parsed = parseWbaUri(wbaUri);
  return resolveHandle(parsed.handle, options);
}

function stripWbaScheme(handleOrUri: string): string {
  return handleOrUri.startsWith('wba://') ? handleOrUri.slice('wba://'.length) : handleOrUri;
}

function normalizeStatus(value: string): HandleStatus {
  switch (value.toLowerCase()) {
    case HandleStatus.Active:
      return HandleStatus.Active;
    case HandleStatus.Suspended:
      return HandleStatus.Suspended;
    case HandleStatus.Revoked:
      return HandleStatus.Revoked;
    default:
      throw new HandleResolutionError(`Unexpected handle status '${value}'`, 502);
  }
}

function normalizeProfile(value: unknown): DIDSubjectProfile | undefined {
  if (!isRecord(value)) {
    return undefined;
  }
  return {
    type: value.type ? String(value.type) : undefined,
    subject_did: String(value.subject_did ?? ''),
    subject_type: normalizeSubjectType(value.subject_type),
    handle: value.handle ? String(value.handle) : undefined,
    display_name: value.display_name ? String(value.display_name) : undefined,
    description: value.description ? String(value.description) : undefined,
    avatar_uri: value.avatar_uri ? String(value.avatar_uri) : undefined,
    profile_uri: value.profile_uri ? String(value.profile_uri) : undefined,
    discoverability: value.discoverability ? String(value.discoverability) : undefined,
    labels: isRecord(value.labels) ? value.labels : undefined,
    updated: value.updated ? String(value.updated) : undefined,
    versionId: value.versionId ? String(value.versionId) : undefined,
    ttl: typeof value.ttl === 'number' ? value.ttl : undefined,
    proof: isRecord(value.proof) ? value.proof : undefined,
  };
}

function dropInvalidProfileProjection(document: HandleResolutionDocument): void {
  if (!document.profile) {
    return;
  }
  if (document.profile.subject_did !== document.did) {
    document.profile = undefined;
    return;
  }
  if (document.profile.handle && document.profile.handle !== document.handle) {
    document.profile = undefined;
  }
}

function normalizeSubjectType(value: unknown): SubjectType {
  if (!value) {
    return SubjectType.Unknown;
  }
  const normalized = String(value).toLowerCase();
  return Object.values(SubjectType).includes(normalized as SubjectType)
    ? (normalized as SubjectType)
    : SubjectType.Unknown;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

File: rust/src/wns/resolver.rs
#[cfg(feature = "network")]
use reqwest::Client;
#[cfg(feature = "network")]
use serde_json::Value;

use super::errors::HandleResolutionError;
#[cfg(feature = "network")]
use super::errors::{HandleGoneError, HandleMovedError, HandleNotFoundError};
use super::models::HandleResolutionDocument;
use super::validator::{build_resolution_url, parse_wba_uri, validate_handle};

#[derive(Debug, Clone)]
pub struct ResolveHandleOptions {
    pub timeout_seconds: f64,
    pub verify_ssl: bool,
    pub base_url_override: Option<String>,
}

impl Default for ResolveHandleOptions {
    fn default() -> Self {
        Self {
            timeout_seconds: 10.0,
            verify_ssl: true,
            base_url_override: None,
        }
    }
}

pub async fn resolve_handle(
    handle: &str,
) -> Result<HandleResolutionDocument, HandleResolutionError> {
    resolve_handle_with_options(handle, &ResolveHandleOptions::default()).await
}

pub async fn resolve_handle_with_options(
    handle: &str,
    options: &ResolveHandleOptions,
) -> Result<HandleResolutionDocument, HandleResolutionError> {
    let bare_handle = strip_wba_scheme(handle);
    let (local_part, domain) =
        validate_handle(bare_handle).map_err(|err| HandleResolutionError {
            message: err.message,
            status_code: 400,
        })?;
    let url = options
        .base_url_override
        .as_ref()
        .map(|base| {
            format!(
                "{}/.well-known/handle/{}",
                base.trim_end_matches('/'),
                local_part
            )
        })
        .unwrap_or_else(|| build_resolution_url(&local_part, &domain));
    let normalized = format!("{}.{}", local_part, domain);

    #[cfg(not(feature = "network"))]
    {
        let _ = url;
        return Err(HandleResolutionError {
            message: format!("Network support is disabled for handle '{}'", normalized),
            status_code: 502,
        });
    }

    #[cfg(feature = "network")]
    {
        let client = Client::builder()
            .danger_accept_invalid_certs(!options.verify_ssl)
            .timeout(std::time::Duration::from_secs_f64(options.timeout_seconds))
            .build()
            .map_err(|_| HandleResolutionError {
                message: format!("Unexpected error resolving handle '{}'", normalized),
                status_code: 502,
            })?;
        let response = client
            .get(url)
            .header("Accept", "application/json")
            .send()
            .await
            .map_err(|err| HandleResolutionError {
                message: format!("Network error resolving handle '{}': {}", normalized, err),
                status_code: 502,
            })?;

        let status = response.status();
        if status.as_u16() == 301 {
            let redirect_url = response
                .headers()
                .get("Location")
                .and_then(|value| value.to_str().ok())
                .unwrap_or_default()
                .to_string();
            return Err(HandleResolutionError {
                message: HandleMovedError {
                    message: format!("Handle '{}' has been migrated", normalized),
                    status_code: 301,
                    redirect_url,
                }
                .message,
                status_code: 301,
            });
        }
        if status.as_u16() == 404 {
            return Err(HandleResolutionError {
                message: HandleNotFoundError {
                    message: format!("Handle '{}' does not exist", normalized),
                    status_code: 404,
                }
                .message,
                status_code: 404,
            });
        }
        if status.as_u16() == 410 {
            return Err(HandleResolutionError {
                message: HandleGoneError {
                    message: format!("Handle '{}' has been permanently revoked", normalized),
                    status_code: 410,
                }
                .message,
                status_code: 410,
            });
        }
        if status.as_u16() != 200 {
            let text = response.text().await.unwrap_or_default();
            return Err(HandleResolutionError {
                message: format!(
                    "Unexpected status {} resolving '{}': {}",
                    status.as_u16(),
                    normalized,
                    text
                ),
                status_code: 502,
            });
        }

        let data: Value = response.json().await.map_err(|err| HandleResolutionError {
            message: format!(
                "Unexpected error resolving handle '{}': {}",
                normalized, err
            ),
            status_code: 502,
        })?;
        let mut document: HandleResolutionDocument =
            serde_json::from_value(data).map_err(|err| HandleResolutionError {
                message: format!(
                    "Unexpected error resolving handle '{}': {}",
                    normalized, err
                ),
                status_code: 502,
            })?;
        if document.handle.to_ascii_lowercase() != normalized {
            return Err(HandleResolutionError {
                message: format!(
                    "Handle mismatch: requested '{}', got '{}'",
                    normalized, document.handle
                ),
                status_code: 502,
            });
        }
        document.drop_invalid_profile_projection();
        Ok(document)
    }
}

pub fn resolve_handle_sync(
    handle: &str,
) -> Result<HandleResolutionDocument, HandleResolutionError> {
    let runtime = tokio::runtime::Runtime::new().map_err(|_| HandleResolutionError {
        message: "Unable to start runtime".to_string(),
        status_code: 502,
    })?;
    runtime.block_on(resolve_handle(handle))
}

pub async fn resolve_handle_from_uri(
    wba_uri: &str,
) -> Result<HandleResolutionDocument, HandleResolutionError> {
    let parsed = parse_wba_uri(wba_uri).map_err(|err| HandleResolutionError {
        message: err.message,
        status_code: 400,
    })?;
    resolve_handle(&parsed.handle).await
}

fn strip_wba_scheme(handle_or_uri: &str) -> &str {
    handle_or_uri
        .strip_prefix("wba://")
        .unwrap_or(handle_or_uri)
}
